packages/gatsby/src/utils/page-ssr-module/lambda.ts

```diff
@@ -289,7 +302,69 @@ async function getEngine(): Promise<GraphQLEngineType> {
   return graphqlEngine
 }
 
-const engineReadyPromise = getEngine()
+let memoizedGraphqlEnginePromise: Promise<GraphQLEngineType> | null = null
+const originToGraphqlEnginePromise = new Map<
+  string,
+  Promise<GraphQLEngineType> | null | Error
+>()
+
+function tryToDownloadEngineFromCollectedOrigins(): Promise<GraphQLEngineType> {
+  for (const [origin, originEngineState] of originToGraphqlEnginePromise) {
+    if (!(originEngineState instanceof Error)) {
+      if (originEngineState === null) {
+        const engineForOriginPromise = getGraphqlEngineInner(origin).catch(
+          e => {
+            originToGraphqlEnginePromise.set(
+              origin,
+              e instanceof Error ? e : new Error(e)
+            )
+
+            if (e.downloadError) {
+              return tryToDownloadEngineFromCollectedOrigins()
+            }
+
+            throw e
+          }
+        )
+        originToGraphqlEnginePromise.set(origin, engineForOriginPromise)
+        return engineForOriginPromise
+      } else {
+        return originEngineState
+      }
+    }
+  }
+
+  return Promise.reject(new Error(`No engine available`))
+}
+
+function getGraphqlEngine(
+  req?: GatsbyFunctionRequest
+): Promise<GraphQLEngineType> {
+  const origin = req?.rawUrl ? new URL(req.rawUrl).origin : cdnDatastoreOrigin
```

Comment:
`origin` tracking here is to allow "retrying" to download datastore in case URL we pass from build wouldn't work (this is the case for CLI deploys which we use in our e2e tests, where that URL is wrong).

We still attempt to start downloading datastore at function init time as this is best perf/timing wise, but if that wouldn't work we will retry with URLs derived from "first request" url.

This is both useful "bug fix" but also requierment to test changes here in our e2e as otherwise we couldn't test case of excluding datastore from functions (one of test variants now exclude datastore)